Sources/Contentful/ArrayResponse.swift

```diff
@@ -95,62 +96,75 @@ public struct ArrayResponse<ItemType>: HomogeneousArray where ItemType: Resource
     }
 }
 
-
 extension ArrayResponse: Decodable {
     public init(from decoder: Decoder) throws {
         let container   = try decoder.container(keyedBy: ArrayCodingKeys.self)
-        items           = try container.decode([ItemType].self, forKey: .items)
-        includes        = try container.decodeIfPresent(ArrayResponse.Includes.self, forKey: .includes)
+
         skip            = try container.decode(UInt.self, forKey: .skip)
         total           = try container.decode(UInt.self, forKey: .total)
         limit           = try container.decode(UInt.self, forKey: .limit)
         errors          = try container.decodeIfPresent([ArrayResponseError].self, forKey: .errors)
 
-        // Workaround for type system not allowing cast of items to [Entry]
-        let entries: [Entry] = items.compactMap { $0 as? Entry }
+        // First see if we can decode an array of user-defined types.
+        if ItemType.self is EntryDecodable.Type {
 
-        let allIncludedEntries = entries + (includedEntries ?? [])
+            // All items and includes.
+            includes = nil
+            mappedIncludes       = try container.decodeIfPresent(MappedIncludes.self, forKey: .includes)
 
-        // Rememember `Entry`s are classes (passed by reference) so we can change them in place
-        for entry in allIncludedEntries {
-            entry.resolveLinks(against: allIncludedEntries, and: (includedAssets ?? []))
-        }
-    }
-    fileprivate enum CodingKeys: String, CodingKey {
-        case items, includes, skip, limit, total, errors
-    }
-}
+            // A copy as an array of dictionaries just to extract "sys.type" field.
+            guard let jsonItems = try container.decode(Swift.Array<Any>.self, forKey: .items) as? [[String: Any]] else {
+                throw SDKError.unparseableJSON(data: nil, errorMessage: "SDK was unable to serialize returned resources")
+            }
+            var entriesJSONContainer = try container.nestedUnkeyedContainer(forKey: .items)
+            var entries: [EntryDecodable] = []
+            let contentTypes = decoder.userInfo[.contentTypesContextKey] as! [ContentTypeId: EntryDecodable.Type]
+
+            while entriesJSONContainer.isAtEnd == false {
```

Comment:
Can't you do something like `while !entriesJSONContainer.isAtEnd` instead of `== false` here? Though I'm not sure what's the most swift-y idiom.

Reply:
For me, it's more about how it would read in english. I think in this case, your suggestion reads better,

"while the container is not at the end" sounds better than 'while the container at the end is false"

I'm happy to change.